File: party/ruby/service.rb
require 'rubygems'
require 'uri'
require 'builder'
require 'rexml/document'
require 'sinatra/base'

module Party

    class Service < Sinatra::Base

        get '/' do

            redirect url('/index.html')

        end

        get '/party/byname/:name' do

            res = THE_DB.get_party_byname(params[:name])

            builder :party_byname, :locals => { :parties => res }

        end


        get '/party/:id' do

            res = THE_DB.get_party(params[:id])
            
            not_found if res == nil

            party_type = res['type']

            builder :party, :locals => { :party => res }

        end    

        get '/party/:id/contacts' do

             res = THE_DB.get_contact_bypartyid(params[:id])

             builder :party_contacts, :locals => { :contacts => res }

        end

        get '/party/:party_id/contact/:contact_id' do

            party_id = params[:party_id]
            contact_id = params[:contact_id]

            res = THE_DB.get_party_contact(party_id, contact_id)

            builder :party_contact, :locals => { :party_contact => res }

        end

        post '/party/:party_id/contact/:contact_id' do

            xml = REXML::Document.new(request.body.read)

            party_id = params[:party_id].to_i
            contact_id = params[:contact_id].to_i
            valid_from = xml.get_text('/PartyContact/ValidFrom')
            valid_until = xml.get_text('/PartyContact/ValidUntil')

            if THE_DB.get_party_contact(party_id, contact_id)
                THE_DB.update_party_contact(party_id, contact_id, valid_from, valid_until)
            else
                THE_DB.create_party_contact(party_id, contact_id, valid_from, valid_until)
            end

            redirect url("/party/#{party_id}/contact/#{contact_id}")

        end

        post '/party/:party_id/contact/:contact_id/delete' do

            party_id = params[:party_id]
            contact_id = params[:contact_id]

            not_found unless THE_DB.get_party_contact(party_id, contact_id)

            THE_DB.delete_party_contact(party_id, contact_id)

            builder :acknowledge, :locals => { :message => "Contact #{contact_id} removed from party #{party_id}" }

        end

        post '/person' do
            
            xml = REXML::Document.new(request.body.read)

            first_name = xml.get_text('/Person/FirstName')
            last_name = xml.get_text('/Person/LastName')
            date_of_birth = xml.get_text('/Person/DateOfBirth')

            party_id = THE_DB.create_person(first_name, last_name, date_of_birth)

            if party_id
                redirect url("/party/#{party_id}")
            end

        end

        post '/person/:id' do

            party_id = params[:id]

            xml = REXML::Document.new(request.body.read)

            first_name = xml.get_text('/Person/FirstName')
            last_name = xml.get_text('/Person/LastName')
            date_of_birth = xml.get_text('/Person/DateOfBirth')

            pty = THE_DB.get_party(party_id)
            not_found unless pty
            raise "Party #{party_id} is not a person" unless pty['type'] == 'P'
            THE_DB.update_person(party_id, first_name, last_name, date_of_birth)

            redirect url("/party/#{party_id}")

        end

        post '/business' do

            xml = REXML::Document.new(request.body.read)

            name = xml.get_text('/Business/Name')

            party_id = THE_DB.create_business(name)

            if party_id
                redirect url("/party/#{party_id}")
            end

        end

        post '/business/:id' do

            party_id = params[:id]

            xml = REXML::Document.new(request.body.read)

            name = xml.get_text('/Business/Name')

            pty = THE_DB.get_party(party_id)
            not_found unless pty
            raise "Party #{party_id} is not a business" unless pty['type'] == 'B'
            THE_DB.update_business(party_id, name)

            redirect url("/party/#{party_id}")

        end

        post '/party/:id/delete' do
            
            not_found unless THE_DB.get_party(params[:id])

            THE_DB.delete_party(params[:id])

            builder :acknowledge, :locals => { :message => "Party #{params[:id]} deleted" }

        end


        not_found do
           
           status 404

           builder :error, :locals => { :name => nil, :message => 'Resource not found' }

        end

        error do

           the_error = env['sinatra.error']

           name = nil
           message = 'Unknown error'

           if the_error.respond_to? :name
               name = the_error.name
           end
          
           if the_error.respond_to? :message
               message = the_error.message
           else
               messsage = the_error
           end

           builder :error, :locals => { :name => name, :message => message }

        end

        get '/contact/:id' do 

            res = THE_DB.get_contact(params[:id])

            not_found if res == nil

            builder :contact, :locals => { :contact => res }

        end

        post '/address' do

            xml = REXML::Document.new(params.to_s)

            contact_id = xml.get_text('/Address/@Id')
            street = xml.get_text('/Address/Street')
            town = xml.get_text('/Address/Town')
            county = xml.get_text('/Address/County')
            post_code = xml.get_text('/Address/PostCode')

            if contact_id
                ct = THE_DB.get_contact(contact_id)
                not_found unless ct
                raise "Contact #{contact_id} is not an address" unless ct['type'] == 'A'
                THE_DB.update_address(contact_id, street, town, county, post_code)
            else
                contact_id = THE_DB.create_address(street, town, county, post_code)
            end

            if contact_id
                redirect url("/contact/#{contact_id}")
            end

        end

        post '/telephone' do

            xml = REXML::Document.new(request.body.read)

            number = xml.get_text('/Telephone/Number')
            type = xml.get_text('/Telephone/SubType')

            contact_id = THE_DB.create_telephone(number, type)

            if contact_id
                redirect url("/contact/#{contact_id}")
            end

        end

        post '/telephone/:id' do

            contact_id = params[:id]
            xml = REXML::Document.new(request.body.read)

            number = xml.get_text('/Telephone/Number')
            type = xml.get_text('/Telephone/SubType')

            ct = THE_DB.get_contact(contact_id)
            not_found unless ct
            raise "Contact #{contact_id} is not a telephone" unless ct['type'] == 'T'
            THE_DB.update_telephone(contact_id, number, type)

            redirect url("/contact/#{contact_id}")

        end

        post '/email' do

            xml = REXML::Document.new(request.body.read)

            address = xml.get_text('/Email/Address')
            type = xml.get_text('/Email/SubType')

            contact_id = THE_DB.create_email(address, type)

            if contact_id
                redirect url("/contact/#{contact_id}")
            end

        end

        post '/email/:id' do

            xml = REXML::Document.new(request.body.read)

            contact_id = params[:id]

            address = xml.get_text('/Email/Address')
            type = xml.get_text('/Email/SubType')

            ct = THE_DB.get_contact(contact_id)
            not_found unless ct
            raise "Contact #{contact_id} is not an email" unless ct['type'] == 'E'
            THE_DB.update_email(contact_id, address, type)
            redirect url("/contact/#{contact_id}")

        end


        post '/contact/:id/delete' do
            
            not_found unless THE_DB.get_contact(params[:id])

            THE_DB.delete_contact(params[:id])

            builder :acknowledge, :locals => { :message => "Contact #{params[:id]} deleted" }

        end


        run! if app_file == $0

    end

end

File: party/dotnet/PartyService/Service.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Net;
using System.ServiceModel;
using System.ServiceModel.Web;
using System.IO;


namespace PartyService
{

    public class Service : IService, IStaticItemService
    {
        private IDatabase database;

        public Service()
        {
            string dbClassName = Properties.Settings.Default.DatabaseClass;
            string dbConnectionString = Properties.Settings.Default.DatabaseConnectionString;

            if (dbClassName == String.Empty || dbConnectionString == string.Empty)
            {
                throw (new Exception("Service Configuration Error"));
            }
            database = (IDatabase)Activator.CreateInstance(Type.GetType(dbClassName));
            database.setConnectionString(dbConnectionString);

        }

        public Party GetParty(string idStr)
        {

            try
            {
                int id;

                if (!int.TryParse(idStr, out id))
                {
                    throw (new WebFaultException<Error>(
                        new Error(string.Format("'{0}' is not a valid party identifier", idStr)),
                        HttpStatusCode.BadRequest));
                }

                Party p = database.GetParty(id);

                if (p == null)
                {
                    throw (new WebFaultException<Error>(
                        new Error(string.Format("Party {0} not found", id)),
                        HttpStatusCode.NotFound));
                }

                return p;
            }
            catch (Exception ex)
            {
                throw MakeWebFaultException(ex);
            }

        }

        public PartyResults GetPartiesByName(string name)
        {
            try
            {
                string baseUrl = WebOperationContext.Current.IncomingRequest.UriTemplateMatch.BaseUri.AbsoluteUri;

                PartyResults partyResults = database.GetPartyByName(name, baseUrl);

                return partyResults;

            }
            catch (Exception ex)
            {
                throw MakeWebFaultException(ex);
            }
        }

        public Person CreatePerson(Person person)
        {
            try
            {
                int newId = database.CreatePerson(person.FirstName, person.LastName, person.DateOfBirth);
                return (Person)GetParty(Convert.ToString(newId));
            }
            catch (Exception ex)
            {
                throw MakeWebFaultException(ex);
            }
        }

        public Person UpdatePerson(string idStr, Person person)
        {
            try
            {
                int id;

                if (!int.TryParse(idStr, out id))
                {
                    throw (new WebFaultException<Error>(
                        new Error(string.Format("'{0}' is not a valid party identifier", idStr)),
                        HttpStatusCode.BadRequest));
                }

                if (database.GetParty(id) == null)
                {
                    throw (new WebFaultException<Error>(
                        new Error(string.Format("Person {0} not found", id)),
                        HttpStatusCode.NotFound));
                }

                database.UpdatePerson(id, person.FirstName, person.LastName, person.DateOfBirth);

                return (Person)GetParty(idStr);

            }
            catch (Exception ex)
            {
                throw MakeWebFaultException(ex);
            }
        }

        public Business CreateBusiness(Business business)
        {
            try
            {
                int newId = database.CreateBusiness(business.Name, business.RegNumber);
                return (Business)GetParty(Convert.ToString(newId));
            }
            catch (Exception ex)
            {
                throw MakeWebFaultException(ex);
            }
        }

        public Business UpdateBusiness(string idStr, Business business)
        {
            try
            {
                int id;

                if (!int.TryParse(idStr, out id))
                {
                    throw (new WebFaultException<Error>(
                        new Error(string.Format("'{0}' is not a valid party identifier", idStr)),
                        HttpStatusCode.BadRequest));
                }

                if (database.GetParty(id) == null)
                {
                    throw (new WebFaultException<Error>(
                        new Error(string.Format("Business {0} not found", id)),
                        HttpStatusCode.NotFound));
                }

                database.UpdateBusiness(id, business.Name, business.RegNumber);

                return (Business)GetParty(idStr);
            }
            catch (Exception ex)
            {
                throw MakeWebFaultException(ex);
            }

        }

        public Acknowledgement DeleteParty(string idStr)
        {

            try
            {
                int id;

                if (!int.TryParse(idStr, out id))
                {
                    throw (new WebFaultException<Error>(
                        new Error(string.Format("'{0}' is not a valid party identifier", idStr)),
                        HttpStatusCode.BadRequest));
                }

                if (database.GetParty(id) == null)
                {
                    throw (new Exception(string.Format("Party {0} not found", id)));
                }

                database.DeleteParty(id);

                return new Acknowledgement(
                    string.Format("Party {0} deleted", id));
            }
            catch (Exception ex)
            {
                throw MakeWebFaultException(ex);
            }

        }

        public Contact GetContact(string idStr)
        {
            try
            {
                int id;

                if (!int.TryParse(idStr, out id))
                {
                    throw (new WebFaultException<Error>(
                        new Error(string.Format("'{0}' is not a valid contact identifier", idStr)),
                        HttpStatusCode.BadRequest));
                }

                Contact c = database.GetContact(id);

                if (c == null)
                {
                    throw (new WebFaultException<Error>(
                        new Error(string.Format("Contact {0} not found", id)),
                        HttpStatusCode.NotFound));
                }

                return c;

            }
            catch (Exception ex)
            {
                throw MakeWebFaultException(ex);
            }
        }

        public Address CreateAddress(Address address)
        {
            try
            {
                int newId = database.CreateAddress(address.Street, address.Town, address.County, address.PostCode);
                return (Address)GetContact(Convert.ToString(newId));
            }
            catch (Exception ex)
            {
                throw MakeWebFaultException(ex);
            }
        }

        public Address UpdateAddress(string idStr, Address address)
        {
            try
            {
                int id;

                if (!int.TryParse(idStr, out id))
                {
                    throw (new WebFaultException<Error>(
                        new Error(string.Format("'{0}' is not a valid contact identifier", idStr)),
                        HttpStatusCode.BadRequest));
                }

                if (database.GetContact(id) == null)
                {
                    throw (new WebFaultException<Error>(
                        new Error(string.Format("Address {0} not found", id)),
                        HttpStatusCode.NotFound));
                }

                database.UpdateAddress(id, address.Street, address.Town, address.County, address.PostCode);

                return (Address)GetContact(idStr);
            }
            catch (Exception ex)
            {
                throw MakeWebFaultException(ex);
            }
        }

        public Email CreateEmail(Email email)
        {
            try
            {
                int newId = database.CreateEmail(email.SubType, email.Address);
                return (Email)GetContact(Convert.ToString(newId));
            }
            catch (Exception ex)
            {
                throw MakeWebFaultException(ex);
            }
        }

        public Email UpdateEmail(string idStr, Email email)
        {
            try
            {
                int id;

                if (!int.TryParse(idStr, out id))
                {
                    throw (new WebFaultException<Error>(
                        new Error(string.Format("'{0}' is not a valid contact identifier", idStr)),
                        HttpStatusCode.BadRequest));
                }

                if (database.GetContact(id) == null)
                {
                    throw (new WebFaultException<Error>(
                        new Error(string.Format("Email {0} not found", id)),
                        HttpStatusCode.NotFound));
                }

                database.UpdateEmail(id, email.SubType, email.Address);

                return (Email)GetContact(idStr);
            }
            catch (Exception ex)
            {
                throw MakeWebFaultException(ex);
            }
        }

        public Telephone CreateTelephone(Telephone telephone)
        {
            try
            {
                int newId = database.CreateTelephone(telephone.SubType, telephone.Number);
                return (Telephone)GetContact(Convert.ToString(newId));
            }
            catch (Exception ex)
            {
                throw MakeWebFaultException(ex);
            }
        }

        public Telephone UpdateTelephone(string idStr, Telephone telephone)
        {
            try
            {
                int id;

                if (!int.TryParse(idStr, out id))
                {
                    throw (new WebFaultException<Error>(
                        new Error(string.Format("'{0}' is not a valid contact identifier", idStr)),
                        HttpStatusCode.BadRequest));
                }

                if (database.GetContact(id) == null)
                {
                    throw (new WebFaultException<Error>(
                        new Error(string.Format("Telephone {0} not found", id)),
                        HttpStatusCode.NotFound));
                }

                database.UpdateTelephone(id, telephone.SubType, telephone.Number);

                return (Telephone)GetContact(idStr);
            }
            catch (Exception ex)
            {
                throw MakeWebFaultException(ex);
            }
        }

        public Acknowledgement DeleteContact(string idStr)
        {

            try
            {
                int id;

                if (!int.TryParse(idStr, out id))
                {
                    throw (new WebFaultException<Error>(
                        new Error(string.Format("'{0}' is not a valid contact identifier", idStr)),
                        HttpStatusCode.BadRequest));
                }

                if (database.GetContact(id) == null)
                {
                    throw (new Exception(string.Format("Contact {0} not found", id)));
                }

                database.DeleteContact(id);

                return new Acknowledgement(
                    string.Format("Contact {0} deleted", id));
            }
            catch (Exception ex)
            {
                throw MakeWebFaultException(ex);
            }

        }

        public PartyContact GetPartyContact(string partyIdStr, string contactIdStr)
        {

            try
            {

                int partyId, contactId;

                if (!int.TryParse(partyIdStr, out partyId))
                {
                    throw (new WebFaultException<Error>(
                        new Error(string.Format("'{0}' is not a valid party identifier", partyIdStr)),
                        HttpStatusCode.BadRequest));
                }

                if (!int.TryParse(contactIdStr, out contactId))
                {
                    throw (new WebFaultException<Error>(
                        new Error(string.Format("'{0}' is not a valid contact identifier", contactIdStr)),
                        HttpStatusCode.BadRequest));
                }

                PartyContact partyContact = database.GetPartyContact(partyId, contactId);

                if (partyContact == null)
                {
                    throw (new WebFaultException<Error>(
                        new Error(string.Format("Party {0} is not linked to contact {1}", partyId, contactId)),
                        HttpStatusCode.NotFound));
                }

                return partyContact;

            }
            catch (Exception ex)
            {
                throw MakeWebFaultException(ex);
            }

        }

        public PartyContact UpdatePartyContact(string partyIdStr, string contactIdStr, PartyContact partyContact)
        {

            try
            {

                int partyId, contactId;

                if (!int.TryParse(partyIdStr, out partyId))
                {
                    throw (new WebFaultException<Error>(
                        new Error(string.Format("'{0}' is not a valid party identifier", partyIdStr)),
                        HttpStatusCode.BadRequest));
                }

                if (!int.TryParse(contactIdStr, out contactId))
                {
                    throw (new WebFaultException<Error>(
                        new Error(string.Format("'{0}' is not a valid contact identifier", contactIdStr)),
                        HttpStatusCode.BadRequest));
                }

                PartyContact pc = database.GetPartyContact(partyId, contactId);

                if (pc == null)
                {
                    database.CreatePartyContact(partyId, contactId, partyContact.ValidFrom, partyContact.ValidUntil);
                }
                else
                {
                    database.UpdatePartyContact(partyId, contactId, partyContact.ValidFrom, partyContact.ValidUntil);
                }
                return GetPartyContact(partyIdStr, contactIdStr);
            }
            catch (Exception ex)
            {
                throw MakeWebFaultException(ex);
            }

        }

        public Acknowledgement DeletePartyContact(string partyIdStr, string contactIdStr)
        {

            try
            {

                int partyId, contactId;

                if (!int.TryParse(partyIdStr, out partyId))
                {
                    throw (new WebFaultException<Error>(
                        new Error(string.Format("'{0}' is not a valid party identifier", partyIdStr)),
                        HttpStatusCode.BadRequest));
                }

                if (!int.TryParse(contactIdStr, out contactId))
                {
                    throw (new WebFaultException<Error>(
                        new Error(string.Format("'{0}' is not a valid contact identifier", contactIdStr)),
                        HttpStatusCode.BadRequest));
                }

                PartyContact pc = database.GetPartyContact(partyId, contactId);

                if (pc == null)
                {
                    throw (new WebFaultException<Error>(
                        new Error(string.Format("Party {0} is not linked to contact {1}", partyId, contactId)),
                        HttpStatusCode.NotFound));
                }

                database.DeletePartyContact(partyId, contactId);

                return new Acknowledgement(
                    string.Format("Contact {0} removed from party {1}", contactId, partyId));
            }
            catch (Exception ex)
            {
                throw MakeWebFaultException(ex);
            }

        }

        public ContactResults GetContactsByParty(string idStr)
        {
            try
            {

                int id;
                string baseUrl = WebOperationContext.Current.IncomingRequest.UriTemplateMatch.BaseUri.AbsoluteUri;

                if (!int.TryParse(idStr, out id))
                {
                    throw (new WebFaultException<Error>(
                        new Error(string.Format("'{0}' is not a valid party identifier", idStr)),
                        HttpStatusCode.BadRequest));
                }

                ContactResults contactResults = database.GetContactByPartyId(id, baseUrl);

                return contactResults;

            }
            catch (Exception ex)
            {
                throw MakeWebFaultException(ex);
            }
        }

        public Stream GetStaticFile(string fileName, string extension)
        {

            fileName.Trim(); extension.Trim();
            fileName.Replace("..", ""); extension.Replace("..", "");

            string contentType = "application/octet-stream";

            switch (extension.ToLower())
            {
                case "xml":
                    contentType = "application/xml";
                    break;
                case "html":
                    contentType = "text/html";
                    break;
                case "txt":
                    contentType = "text/plain";
                    break;
                case "gif":
                    contentType = "image/gif";
                    break;
                case "jpg":
                    contentType = "image/jpg";
                    break;
            }

  
            string staticFileName = string.Format("./public/{0}.{1}", fileName, extension);

            try
            {
                WebOperationContext.Current.OutgoingResponse.ContentType = contentType;
                return File.OpenRead(staticFileName);
            }
            catch
            {
                throw (new WebFaultException<Error>(
                       new Error(string.Format("File access error for '{0}.{1}'", fileName, extension)),
                       HttpStatusCode.NotFound));
            }
          
        }

         private WebFaultException<Error> MakeWebFaultException(Exception ex)
        {
            return (ex is WebFaultException<Error>) ?
                (WebFaultException<Error>)ex :
                new WebFaultException<Error>(
                    new Error(ex.Message), HttpStatusCode.InternalServerError);
        }

    }
}
